feat: add seed phrase backup confirmation during wallet creation

secondl1ght opened the pull request:
As talked about in issue #74 I have added a seed phrase backup confirmation step during the wallet creation process.

![image](https://user-images.githubusercontent.com/85003930/156940812-3df14d61-db3a-4d22-8c8a-1f8b6b9e5afe.png)

![image](https://user-images.githubusercontent.com/85003930/156940828-34ed72c2-18fa-4d50-a1ee-add064fc84e3.png)

![image](https://user-images.githubusercontent.com/85003930/156940839-4cf667c7-fe03-4964-be30-fc4c69cc825b.png)

![image](https://user-images.githubusercontent.com/85003930/156940884-f894ffa5-090b-46d3-a8c7-641edc88fab6.png)

Users can go back to the previous page to write down their seed phrase if they made a mistake, once the seed phrase is confirmed the Back button and textarea is disabled and the Fund Wallet button appears to complete the setup process.

MaxHillebrand commented:
How about doing 12 separate textboxes which are numbered?

And please add a small skip button here.

dnlggr commented:
> How about doing 12 separate textboxes which are numbered?

Agree, let's do it like this!

dnlggr commented:
Btw, [this section](https://bitcoin.design/guide/onboarding/backing-up-a-wallet/) from the Bitcoin Design Guidelines might also be interesting here.

secondl1ght commented:
Ok I have changed it to be 12 word input boxes, I also added a Skip button! It also now gives user feedback as they are typing, so once they type the correct word it will say 'Word matches!' and then they click Verify to do the final confirmation on all words.

New screenshots:

![image](https://user-images.githubusercontent.com/85003930/157083661-645d1ac7-1873-456f-a38d-e579b46a47bd.png)

![image](https://user-images.githubusercontent.com/85003930/157083769-b6aa3ac2-2dab-4436-bd43-7d821800dac0.png)

![image](https://user-images.githubusercontent.com/85003930/157084060-2d4608c9-2edd-4921-b693-fd53597cc545.png)

![image](https://user-images.githubusercontent.com/85003930/157084294-26461e32-a757-416c-b199-41fe0a304d93.png)

MaxHillebrand commented:
Awesome!

Now just consider that there are 12, 16, 18, and 24 word specifications... So there needs to be a way to add more text boxes.

The previous approach nicely catches this, just type in one textbox as many words as you have.

PulpCattel commented:
> Now just consider that there are 12, 16, 18, and 24 word specifications

JoinMarket creates 12 words seed and I don't think there are reasons to change this any time soon, so this shouldn't be a problem.

dnlggr commented on src/components/CreateWallet.jsx:
```diff
@@ -64,47 +64,389 @@ const WalletCreationConfirmation = ({ createdWallet, walletConfirmed }) => {
+          >
+            Back
+          </rb.Button>
+          {seedBackup === false ? (
+            <rb.Button variant="dark" onClick={() => onSubmit()}>
+              Verify
+            </rb.Button>
+          ) : (
+            <rb.Button variant="dark" onClick={() => userConfirmed && walletConfirmed()}>
+              Fund wallet
+            </rb.Button>
+          )}
```
I feel like we could probably go without the extra "verify" step and just show the "fund" button (disabled as long as the entered seedphrase doesn't match, of course).

secondl1ght replied:
Good idea, I've made it smoother and more reactive now.

dnlggr reviewed (commented):
Awesome, thanks @secondl1ght! Works nicely, especially the instant word-by-word feedback is a nice touch! 👌 

I left some feedback inline, let me know if you need anything!

secondl1ght commented:
Thanks for the feedback @dnlggr, it will take me a little while to review all the comments and make the changes. Hoping I have some time and can get it done this week though!

dnlggr commented:
No worries, take your time!

secondl1ght commented on src/components/CreateWallet.jsx:
```diff
@@ -64,47 +64,389 @@ const WalletCreationConfirmation = ({ createdWallet, walletConfirmed }) => {
   const [sensitiveInfoWasRevealed, setSensitiveInfoWasRevealed] = useState(false)
+  const [step, setStep] = useState(0)
+  const [seedBackup, setSeedBackup] = useState(false)
+  const [wordOne, setWordOne] = useState('')
+  const [wordTwo, setWordTwo] = useState('')
+  const [wordThree, setWordThree] = useState('')
+  const [wordFour, setWordFour] = useState('')
+  const [wordFive, setWordFive] = useState('')
+  const [wordSix, setWordSix] = useState('')
+  const [wordSeven, setWordSeven] = useState('')
+  const [wordEight, setWordEight] = useState('')
+  const [wordNine, setWordNine] = useState('')
+  const [wordTen, setWordTen] = useState('')
+  const [wordEleven, setWordEleven] = useState('')
+  const [wordTwelve, setWordTwelve] = useState('')
```
How do you mean use an array for this, do you mind showing an example?

secondl1ght replied:
I am removing this step entirely so I will resolve this comment.

secondl1ght commented on src/components/CreateWallet.jsx:
```diff
@@ -64,47 +64,389 @@ const WalletCreationConfirmation = ({ createdWallet, walletConfirmed }) => {
+                    setWordOne(e.target.value)
+                  }}
+                  disabled={seedBackup}
+                  isInvalid={!wordOne === seedphrase[0]}
```
Good catch thanks!

secondl1ght commented on src/components/CreateWallet.jsx:
```diff
@@ -64,47 +64,389 @@ const WalletCreationConfirmation = ({ createdWallet, walletConfirmed }) => {
+        <rb.Form noValidate>
+          <div className="container">
+            <div className="row mb-4">
+              <div className="col">
+                <rb.InputGroup />
+                <rb.FormControl
+                  type="text"
+                  placeholder="One"
+                  value={wordOne}
+                  onChange={(e) => {
+                    setFeedback(false)
+                    setWordOne(e.target.value)
+                  }}
+                  disabled={seedBackup}
+                  isInvalid={!wordOne === seedphrase[0]}
+                  isValid={wordOne === seedphrase[0]}
+                  required
+                />
+                {wordOne === seedphrase[0] && !seedBackup ? (
+                  <rb.Form.Control.Feedback>Word matches!</rb.Form.Control.Feedback>
+                ) : (
+                  <rb.Form.Control.Feedback type="invalid">Word does not match.</rb.Form.Control.Feedback>
+                )}
+              </div>
```
Is this possible in this situation because every 2 input boxes go inside a container row `div`? So not every input gets the same code if that makes sense.

secondl1ght commented:
Ok I made lots of improvements based on @dnlggr's comments, thanks for the review!

- made a new component for `BackupConfirmation`
- added numbers to each input box so it is more obvious rather than just relying on placeholders
- updated some variable names to be more semantic and shortened some code where possible
- the user now gets feedback while typing for when the word does not match
- the input box becomes disabled when the correct word is entered
- removed the extra 'Verify' step and everything happens reactively now, so as soon as all of the correct words are entered the Fund Wallet button becomes enabled and the other buttons become disabled.

It feels pretty smooth to me now, please have a look and see what you think. There are still a few comments I am working on that I responded to with questions.

New look:

![image](https://user-images.githubusercontent.com/85003930/157350237-8c727f30-2c4a-4cad-9d06-f93337702107.png)

![image](https://user-images.githubusercontent.com/85003930/157350323-043fddbe-d7d9-4236-b324-160f70a7f7ac.png)

![image](https://user-images.githubusercontent.com/85003930/157350447-1732799b-597c-4dc0-a57c-3a81c052a41d.png)

![image](https://user-images.githubusercontent.com/85003930/157350581-e66ec47c-9ac8-440b-8cd8-eabd7efffb3f.png)

secondl1ght commented:
One thing I noticed is that other places in the app use `setAlert` so communicate with the user, do we want to use that for the 'Seed phrase confirmed.' message or do you like the way it is?

dnlggr reviewed (commented):
Nice, the behavior is really good now!

UI wise, I'd say lets see how left aligning the whole block of inputs looks like (since all other content is left aligned as well in this and the previous screen). For the numbers we should probably use the `.slashed-zeroes` style and make sure that the number shave equal width so that the entered seed words don't misalign:

<img width="124" alt="Screenshot 2022-03-09 at 16 02 01" src="https://user-images.githubusercontent.com/10026790/157468443-e1f9f186-ed83-4d29-850a-dbeaccfd7c85.png">

To answer your qustions regarding the array / drying up the repeated code blocks, I pushed a quick example to [confirm-wallet-backup-loop](https://github.com/joinmarket-webui/joinmarket-webui/tree/confirm-wallet-backup-loop). It's not necessarily super clean but definitely shorter and easier to maintain. Maybe use it as guidance for this PR.

It's starting to come together quite nicely, good job. 👏  Let me know if you need anything further!

dnlggr commented on src/components/CreateWallet.jsx:
```diff
@@ -60,53 +60,439 @@ const WalletCreationForm = ({ createWallet, isCreating }) => {
+                  isValid={wordTwo === seedphrase[1]}
+                  required
+                />
+                {wordTwo === seedphrase[1] && !seedBackup && (
+                  <rb.Form.Control.Feedback>Word matches!</rb.Form.Control.Feedback>
+                )}
+                {wordTwo !== seedphrase[1] && wordTwo.length > 0 && (
+                  <rb.Form.Control.Feedback type="invalid">Word does not match.</rb.Form.Control.Feedback>
```
Thinking about it again, we could drop the text feedback. That would prevent the inputs jumping around when feedback appears / disappears. The red / green outline is probably enough of a feedback anyway. Thoughts?

secondl1ght replied:
That would make it even cleaner, will implement.

secondl1ght commented on src/components/CreateWallet.jsx:
```diff
@@ -60,53 +60,439 @@ const WalletCreationForm = ({ createWallet, isCreating }) => {
+        parentStepSetter={childStepSetter}
+        createdWallet={createdWallet}
+        walletConfirmed={walletConfirmed}
+        userConfirmed={userConfirmed}
```
That is true, good point!

secondl1ght commented:
Ok so I have 2 tasks left to do:

- look at your code example @dnlggr for the DRY and implement that
- make sure the numbers are equal width so all words align

secondl1ght commented:
Ok I ported @dnlggr's DRY improvements over and they work nicely, I thought the code works well so I wont try to refactor it any further. I did fix one error where the `key` prop to the `SeedWordInput` component was missing. Also I noticed that you imported `useCallback` but never used it, did you mean to use it somewhere or should I remove the import?

Thanks for doing this refactor, there are lots of things in there that I have not learned yet. I need to get better at using loops to DRY up the code. Some of it is a bit harder to read for me personally but I know it is higher quality code. Once I write things like this a few times it will get easier to read and understand how it all works. I'll use this code to reference in the future when I want to do something similar.

This new backup screen is working really smooth now and looking great, i'll finish the last remaining task they we can get a final review done on it.

secondl1ght commented:
I'm having a hard time getting all of the input numbers to be the same size LOL! Bootstrap is good for some things but I prefer Tailwind, does anyone know the best way to do this? I can't apply styling directly to the RB components right? So I wrapped each one in a `span` element and then set the fixed `col` widths for the `input` and number to left. But when I did that the number become half the height of the input, I dont see anything in the Bootstrap docs about setting the height for an element. I tried doing an inline `style` tag but this didn't work either...

![image](https://user-images.githubusercontent.com/85003930/157588957-87a022bf-fcac-48fb-9011-5915a2bd99ca.png)

dnlggr commented:
Cool thanks for the updates @secondl1ght!

> I did fix one error where the key prop to the SeedWordInput component was missing.

Good catch on the key prop. Is it enough to just pass it to `<SeedWordInput>` and then not use it inside the component? Not sure if `<FormControl>` inside needs a key as well.

> Also I noticed that you imported useCallback but never used it, did you mean to use it somewhere or should I remove the import?

Sure remove the import—it's not needed.

> Some of it is a bit harder to read for me personally but I know it is higher quality code.

No you're right, it's definitely a bit cumbersome to read and could certainly be a lot cleaner. I wouldn't say it's super high quality yet—was just somthing quick to demonstrate how it could work. 😄 I feel like it could probably be cleaner by just using a wrapping flexbox for the seed word inputs (cf. the`Seedphrase` component) and getting rid of the rigid row/col grid. Then we could omit the modulo checks which are a bit awkward. But it's probably fine for now. If you're ok with it I'd say let's keep it like that for now.

>  I can't apply styling directly to the RB components right?

Sure! You can apply style, assign it to classes, everything should work just as with a plain HTML tag. I posted a short snipped to fix the index width (same hack as we use for the `Seedphrase` component).

secondl1ght commented:
Oh wait it looks like that was just an old error I forgot to clear lol, ok ready for final review now. 😄

secondl1ght commented:
> > The key prop was needed because it was causing some errors without it.
> 
> Yeah but the prop is only needed on the array item (the one returned from `map`). You shouldn't need it inside `SeedWordInput`. See [this part](https://reactjs.org/docs/lists-and-keys.html#extracting-components-with-keys) of the React docs for a more detailed explainer.

I tried not including it in `SeedWordInput` and while the app still worked without it, it still logged an error to the console. I figured it would be better to include it so we have no errors even if it still works?

dnlggr commented:
> I tried not including it in SeedWordInput and while the app still worked without it, it still logged an error to the console. I figured it would be better to include it so we have no errors even if it still works?

Can you double check that? I'm not sure this is [how keys work](https://reactjs.org/docs/lists-and-keys.html).

With the current implementation, there's two issues being logged:

> Warning: Each child in a list should have a unique "key" prop.

This refers to the outermost `div`s in the loops _not_ haviong having a key prop.

> Warning: SeedWordInput: `key` is not a prop. Trying to access it will result in `undefined` being returned.

This is because you're passing the key down to the `SeedwordInput`. Passing keys down to components is not needed ([and not supported](https://reactjs.org/warnings/special-props.html) in fact). Keys are only needed for the direct children of a list. You need to set the key prop for [this](https://github.com/joinmarket-webui/joinmarket-webui/pull/156/files#diff-2dc25b9c1ccccb688b2703dc2ad63a02a3732c70d6714b6057a7c80ef226554eR116) and for [this](https://github.com/joinmarket-webui/joinmarket-webui/pull/156/files#diff-2dc25b9c1ccccb688b2703dc2ad63a02a3732c70d6714b6057a7c80ef226554eR120) div—which are the children of a list. `SeedWordInput` doesn't need a key you can remove it from the component completely.

Let me know if you need help, I can't push directly to your branch unfortunately but this is how the loop part should look like:

``` js
        <div className="container slashed-zeroes p-0">
          {[...new Array(seedphrase.length)].map((_, outerIndex) => {
            if (outerIndex % 2 !== 0) return null

            const seedWords = seedphrase.slice(outerIndex, outerIndex + 2)

            return (
              <div className="row mb-4" key={outerIndex}>
                {seedWords.map((seedWord, innerIndex) => {
                  const wordIndex = outerIndex + innerIndex
                  return (
                    <div className="col" key={wordIndex}>
                      <SeedWordInput
                        number={wordIndex + 1}
                        targetWord={seedWord}
                        isValid={seedWordConfirmations[wordIndex]}
                        setIsValid={(isValid) => {
                          setSeedWordConfirmations(
                            seedWordConfirmations.map((confirmation, index) =>
                              index === wordIndex ? isValid : confirmation
                            )
                          )
                        }}
                      />
                    </div>
                  )
                })}
              </div>
            )
          })}
        </div>
```

secondl1ght commented:
Ah ok I see what the issue was for me, those warning only appear on the first render. So as I was updating the code in the dev environment and testing changes live and I thought the errors went away, but if you shut down the entire app and relaunch then the errors show up again.

I have fixed it now and there are no errors (tested with app relaunch), thanks for showing me that! In the future I can push branches directly from this repo instead of my fork so that you have access to it?

I will resolve the merge conflicts now as well, thanks again for all the help on this. 🙌🏼

dnlggr commented:
Awesome thanks for fixing @secondl1ght! 👏 

> It looks like the merge conflicts were the new translations, I resolved what was there but this page will have to be looked at again for new translations. Should I put in a new issue for that?

Just add the new translation directly here that'll be the easiest way to do it imo.

secondl1ght commented:
Ok I have updated the new translations now - this was a big PR for me lol! 🙂

dnlggr reviewed (approved):
Ok looks good to me! ✅  Great job powering through this long PR @secondl1ght! 👏 🏁 👍

secondl1ght commented:
Awesome!! Should I merge this now?? 🙌🏼

dergigi commented:
Thanks again for implementing this @secondl1ght 👏 

I'm not sure if the skip button is such a great idea on mainnet; I tracked an issue here: https://github.com/joinmarket-webui/joinmarket-webui/issues/196
